### PR TITLE
Make Assets Pallet's Privileged Roles Option

### LINKED ISSUE #91: Make Assets Pallet's Privileged Roles `Option`
Change the [privileged asset roles](https://github.com/paritytech/substrate/blob/73c4f94ce0cb41e35bd7fbc7095590b98a351dbb/frame/assets/src/types.rs#L34L41) from `AccountId` to `Option<AccountId>`. When the `owner` is set to `None`, only governance would be able to restore the role.

For discussion, if `owner` is `Some`, and they want to set a lower rank role from `None` to `Some`, should that require governance?

This would look similar to the [`PoolRoles` struct](https://github.com/paritytech/substrate/blob/73c4f94ce0cb41e35bd7fbc7095590b98a351dbb/frame/nomination-pools/src/lib.rs#L528) in Nomination Pools. 

_Originally posted by @joepetrowski in https://github.com/paritytech/substrate/pull/12310#discussion_r982367220_

### LINKED ISSUE #12458: Make Assets Pallet Privileges `Option`
Change the [privileged asset roles](https://github.com/paritytech/substrate/blob/73c4f94ce0cb41e35bd7fbc7095590b98a351dbb/frame/assets/src/types.rs#L34L41) from `AccountId` to `Option<AccountId>`. When the `owner` is set to `None`, only governance would be able to restore 

This would look similar to the [`PoolRoles` struct](https://github.com/paritytech/substrate/blob/73c4f94ce0cb41e35bd7fbc7095590b98a351dbb/frame/nomination-pools/src/lib.rs#L528) in Nomination Pools. 

_Originally posted by @joepetrowski in https://github.com/paritytech/substrate/pull/12310#discussion_r982367220_

### PR DESCRIPTION
**Pull request by dharjeezy**
Include `Option`  for struct `AssetDetails`
Cumulus companion https://github.com/paritytech/cumulus/pull/1912
closes paritytech/polkadot-sdk#91 

Polkadot address: 12ZNas89oEagaxLVNbpqmvfMxdrGrqN7gJKSpwthTUPZsrku

### CODE REVIEW

**Comment by dharjeezy** [comment]
@joepetrowski

**Comment by joepetrowski** [comment]
Thanks! Will need a migration and companion in Cumulus for existing assets.

**Comment by dharjeezy** [comment]
> Thanks! Will need a migration and companion in Cumulus for existing assets.

Oh..Alright

File: frame/assets/src/functions.rs [hunk 1]
```diff
@@ -695,7 +695,7 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
 
 				let metadata = Metadata::<T, I>::take(&id);
 				T::Currency::unreserve(
-					&details.owner,
+					&details.owner.unwrap(),
```
[the remark this answers is not in the document]
Comment (joepetrowski) [reply frame/assets/src/functions.rs:698]: Good point. Maybe then we need to track a `depositor` for the class?

[the remark this answers is not in the document]
Comment (jsidorenko) [reply frame/assets/src/functions.rs:698]: You would need to do a bunch of actions:
1) call the unreserve() method if the owner is set
2) don't apply(take) the deposit when the owner is None
3) when you change the `owner` - return or repatriate the deposit

[the remark this answers is not in the document]
Comment (dharjeezy) [reply frame/assets/src/functions.rs]: Alright @jsidorenko

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2080249

File: frame/assets/src/impl_fungibles.rs [hunk 2]
```diff
@@ -268,18 +268,18 @@ impl<T: Config<I>, I: 'static> fungibles::roles::Inspect<<T as SystemConfig>::Ac
 	for Pallet<T, I>
 {
 	fn owner(asset: T::AssetId) -> Option<<T as SystemConfig>::AccountId> {
-		Asset::<T, I>::get(asset).map(|x| x.owner)
+		Asset::<T, I>::get(asset).map(|x| x.owner.unwrap())
```
[the remark this answers is not in the document]
Comment (dharjeezy) [reply frame/assets/src/impl_fungibles.rs:273]: Been wrapping my head around this.
What do you suggest in this case?
 @jsidorenko

File: frame/assets/src/functions.rs [hunk 3]
```diff
@@ -677,7 +677,7 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
 		Asset::<T, I>::try_mutate_exists(id, |maybe_details| -> Result<(), DispatchError> {
 			let mut details = maybe_details.as_mut().ok_or(Error::<T, I>::Unknown)?;
 			if let Some(check_owner) = maybe_check_owner {
```
Comment (jsidorenko) [inline frame/assets/src/functions.rs:679]: Same as above:
```rust
			if maybe_check_owner.is_some() {
				ensure!(maybe_check_owner == details.owner, Error::<T, I>::NoPermission);
			}
```

File: frame/assets/src/functions.rs [hunk 4]
```diff
@@ -770,10 +770,12 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
 			ensure!(details.approvals == 0, Error::<T, I>::InUse);
 
 			let metadata = Metadata::<T, I>::take(&id);
-			T::Currency::unreserve(
-				&details.owner,
-				details.deposit.saturating_add(metadata.deposit),
-			);
+			if details.owner.is_some() {
```
Comment (jsidorenko) [inline frame/assets/src/functions.rs:773]: pls replace with `if let Some() = details.owner`

File: frame/assets/src/lib.rs [hunk 5]
```diff
@@ -874,7 +874,7 @@ pub mod pallet {
 				d.status == AssetStatus::Live || d.status == AssetStatus::Frozen,
 				Error::<T, I>::AssetNotLive
 			);
-			ensure!(origin == d.freezer, Error::<T, I>::NoPermission);
+			ensure!(origin == d.freezer.unwrap(), Error::<T, I>::NoPermission);
```
Comment (jsidorenko) [inline frame/assets/src/lib.rs:877]: pls go through all the places where you use `.unwrap()` and refactor them

File: frame/assets/src/types.rs [hunk 6]
```diff
@@ -32,7 +32,7 @@ pub(super) type AssetAccountOf<T, I> =
 /// AssetStatus holds the current state of the asset. It could either be Live and available for use,
 /// or in a Destroying state.
 #[derive(Clone, Encode, Decode, Eq, PartialEq, RuntimeDebug, MaxEncodedLen, TypeInfo)]
-pub(super) enum AssetStatus {
+pub enum AssetStatus {
```
Comment (jsidorenko) [inline frame/assets/src/types.rs:35]: why did you remove the `(super)`?

**Review by jsidorenko — CHANGES_REQUESTED** [review]
Pls check the comments above

File: frame/assets/src/types.rs (hunk 6 above)
Comment (dharjeezy) [reply frame/assets/src/types.rs:35]: I had to remove the (super) because of the runtime migration

**Comment by dharjeezy** [comment]
> Pls check the comments above

Alright, Will do as suggested

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2175276

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2175275

File: frame/assets/src/functions.rs [hunk 7]
```diff
@@ -564,7 +564,7 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
 
 			// Check admin rights.
 			if let Some(need_admin) = maybe_need_admin {
```
Comment (jsidorenko) [inline frame/assets/src/functions.rs:566]: Pls rewrite this in the same way as above

Comment (dharjeezy) [reply frame/assets/src/functions.rs]: Done.

**Comment by jsidorenko** [comment]
@joepetrowski I have a bunch of notes regarding the different methods. Pls correct what's right and what's wrong:

- [force_create()](https://github.com/paritytech/substrate/blob/d8a301a553aa15ddef34969111e19935bf7109c0/frame/assets/src/lib.rs#L626) doesn't allow to set the owner as `None`;
- [mint()/burn()/force_transfer()](https://github.com/paritytech/substrate/blob/d8a301a553aa15ddef34969111e19935bf7109c0/frame/assets/src/lib.rs#L741)
are not accessible when the Issuer/Manager/Admin is `None`;
- [freeze()/thaw()](https://github.com/paritytech/substrate/blob/d8a301a553aa15ddef34969111e19935bf7109c0/frame/assets/src/lib.rs#L903) I think those methods should be accessible by the root call as well. Otherwise, it won't be possible to call it at all when the freezer is None;
- [set_team()](https://github.com/paritytech/substrate/blob/d8a301a553aa15ddef34969111e19935bf7109c0/frame/assets/src/lib.rs#L1079) shall it be possible for the user to switch roles from `Some()` to `None` here?
- [force_asset_status()](https://github.com/paritytech/substrate/blob/d8a301a553aa15ddef34969111e19935bf7109c0/frame/assets/src/lib.rs#L1279) shall it be possible to set roles to `None` here? If we switch from `None` to `Some()` do we need to apply the deposit?
- [approve_transfer()](https://github.com/paritytech/substrate/blob/d8a301a553aa15ddef34969111e19935bf7109c0/frame/assets/src/lib.rs#L1336) this feature won't be accessible now if the owner is `None`;

File: frame/assets/src/migration.rs [hunk 8]
```diff
@@ -44,10 +44,10 @@ pub mod v1 {
...
 
 			AssetDetails {
-				owner: self.owner,
-				issuer: self.issuer,
-				admin: self.admin,
-				freezer: self.freezer,
+				owner: Some(self.owner),
+				issuer: Some(self.issuer),
+				admin: Some(self.admin),
+				freezer: Some(self.freezer),
```
Comment (joepetrowski) [inline frame/assets/src/migration.rs:50]: This is not `v1` though. I think [this](https://github.com/paritytech/substrate/blob/3e91823/frame/alliance/src/migration.rs#L26L42) is a better example for how to handle possible multiple migrations.

File: frame/assets/src/migration.rs [hunk 9]
```diff
@@ -120,3 +120,100 @@ pub mod v1 {
...
+				let mut translated = 0u64;
+				Asset::<T>::translate::<
+					OldAssetDetails<T::Balance, T::AccountId, DepositBalanceOf<T>>,
+					_,
+				>(|_key, old_value| {
+					translated.saturating_inc();
+					Some(old_value.migrate_to_v2())
+				});
+				current_version.put::<Pallet<T>>();
+				log::info!(target: "runtime::assets", "Upgraded {} pools, storage to version {:?}", translated, current_version);
```
Comment (joepetrowski) [inline frame/assets/src/migration.rs:178]:
```suggestion
				log::info!(target: "runtime::assets", "Upgraded {} asset classes, storage to version {:?}", translated, current_version);
```

File: frame/assets/src/lib.rs [hunk 10]
```diff
@@ -1040,18 +1041,21 @@ pub mod pallet {
...
 				}
 
 				let metadata_deposit = Metadata::<T, I>::get(id).deposit;
 				let deposit = details.deposit + metadata_deposit;
 
 				// Move the deposit to the new owner.
-				T::Currency::repatriate_reserved(&details.owner, &owner, deposit, Reserved)?;
+				if let Some(details_owner) = &details.owner {
+					T::Currency::repatriate_reserved(details_owner, &owner, deposit, Reserved)?;
+				}
```
Comment (joepetrowski) [inline frame/assets/src/lib.rs:1056]: I think we should _try_ to `unreserve` the amount from the current owner, and then `reserve` it from the new owner. One, this is causing a balance transfer. Two, the deposit costs may have changed since the asset was created. If they have gone up, and the current owner doesn't have enough reserved, it will fail (and the owner will be stuck unless they reserve more via some other operation). Since we've calculated the going rate for deposits here, the new owner should have to reserve the updated amount.

**Comment by joepetrowski** [comment]
> [force_create()](https://github.com/paritytech/substrate/blob/d8a301a553aa15ddef34969111e19935bf7109c0/frame/assets/src/lib.rs#L626) doesn't allow to set the owner as None

I think this should be possible.

> [mint()/burn()/force_transfer()](https://github.com/paritytech/substrate/blob/d8a301a553aa15ddef34969111e19935bf7109c0/frame/assets/src/lib.rs#L741)
are not accessible when the Issuer/Manager/Admin is None;
> [freeze()/thaw()](https://github.com/paritytech/substrate/blob/d8a301a553aa15ddef34969111e19935bf7109c0/frame/assets/src/lib.rs#L903) I think those methods should be accessible by the root call as well. Otherwise, it won't be possible to call it at all when the freezer is None;

I also think these should all be accessibly by `Root` or `ForceOrigin` (which may be Root in the config). In general I think an admin team of `None` is really saying, "only governance should have privileged permission on this asset class".

> [set_team()](https://github.com/paritytech/substrate/blob/d8a301a553aa15ddef34969111e19935bf7109c0/frame/assets/src/lib.rs#L1079) shall it be possible for the user to switch roles from Some() to None here?

Yes, otherwise how will it get into a `None` state?

> [force_asset_status()](https://github.com/paritytech/substrate/blob/d8a301a553aa15ddef34969111e19935bf7109c0/frame/assets/src/lib.rs#L1279) shall it be possible to set roles to None here? If we switch from None to Some() do we need to apply the deposit?

Yeah, `ForceOrigin` should be able to do most operations. I would say yes on applying the deposit.

> [approve_transfer()](https://github.com/paritytech/substrate/blob/d8a301a553aa15ddef34969111e19935bf7109c0/frame/assets/src/lib.rs#L1336) this feature won't be accessible now if the owner is None;

This we definitely need to work, as it gets used by a lot of applications. Ownerless assets shouldn't be less functional at the fungibility level (i.e. the users), only at the management level (i.e. permissioned functions).

**Comment by joepetrowski** [comment]
bot rebase

**Comment by jsidorenko** [comment]
Hi @dharjeezy, what's the progress with this PR? Any plans to finish it soon?

**Comment by dharjeezy** [comment]
> Hey, is anyone still working on this? Due to the inactivity this issue has been automatically marked as stale. It will be closed if no further activity occurs. Thank you for your contributions.

I have been so busy with other things @joepetrowski so sorry for this
i will fix up the migration this week and get this PR done

**Comment by dharjeezy** [comment]
bot rebase

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2567587

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2643196

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2643198

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2643197

**Comment by dharjeezy** [comment]
I have fixed the migration @joepetrowski 
@jsidorenko you can take a look

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2789035

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2790154

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2790386

**Comment by jsidorenko** [comment]
@joepetrowski pls re-open if we still need this
